### PR TITLE
Issue 7633: Add new bucket to ads/rewards p3a metrics.

### PR DESCRIPTION
**Pull request by iefremov**
Fix https://github.com/brave/brave-browser/issues/7633

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/p3a/brave_p3a_service.cc [hunk 1]
```diff
@@ -64,10 +64,10 @@ constexpr const char* kCollectedHistograms[] = {
     "Brave.Core.BookmarksCountOnProfileLoad",
     "Brave.Core.TabCount",
     "Brave.Core.WindowCount",
-    "Brave.Rewards.WalletBalance",
-    "Brave.Rewards.AutoContributionsState",
-    "Brave.Rewards.TipsState",
-    "Brave.Rewards.AdsState",
+    "Brave.Rewards.WalletBalance2",
```
Comment (bsclifton) [inline components/p3a/brave_p3a_service.cc]: can these be in the format like 
```
"Brave.Rewards.WalletBalance.2",
"Brave.Rewards.AutoContributionsState.2",
"Brave.Rewards.TipsState.2",
"Brave.Rewards.AdsState.2",
```
(including references to them above)

Comment (bsclifton) [reply components/p3a/brave_p3a_service.cc]: Also- does it make sense to have them as #defines / constants?

**Comment by iefremov** [comment]
@tmancey Hi Terry, could you please take a look at this, while Nejc is offline?

File: components/brave_rewards/browser/rewards_p3a.cc [hunk 2]
```diff
@@ -34,76 +38,93 @@ void RecordAutoContributionsState(AutoContributionsP3AState state, int count) {
...
       prefs->GetBoolean(brave_rewards::prefs::kBraveRewardsEnabled);
+  const bool ads_enabled = prefs->GetBoolean(brave_ads::prefs::kEnabled);
   if (pref == brave_ads::prefs::kEnabled) {
-    if (prefs->GetBoolean(brave_ads::prefs::kEnabled)) {
+    if (ads_enabled) {
       brave_rewards::RecordAdsState(AdsP3AState::kAdsEnabled);
       prefs->SetBoolean(brave_ads::prefs::kAdsWereDisabled, false);
     } else {
       // Apparently, the pref was disabled.
+      // TODO(ifremov): DCHECK(rewards_enabled)?
```
Comment (tmancey) [inline components/brave_rewards/browser/rewards_p3a.cc]: Was this intentionally left in?

Comment (iefremov) [reply components/brave_rewards/browser/rewards_p3a.cc]: yep, basically I'm not 100% whether it's legal to `DCHECK` here, so any advice is welcome

Comment (tmancey) [reply components/brave_rewards/browser/rewards_p3a.cc]: I think it's a good place to `DCHECK` to make sure rewards is enabled, however, if rewards are not enabled (which should not be the case) we would be incorrectly setting P3A, so maybe a good idea to fail gracefully???

**Comment by iefremov** [comment]
unrelated network audit failure in CI :-/

**Comment by rebron** [comment]
@iefremov network audit failure related to crowd deny?

File: components/brave_rewards/browser/rewards_p3a.cc [hunk 3]
```diff
@@ -34,76 +38,93 @@ void RecordAutoContributionsState(AutoContributionsP3AState state, int count) {
...
       brave_rewards::RecordAdsState(
           rewards_enabled ? AdsP3AState::kAdsEnabledThenDisabledRewardsOn :
                             AdsP3AState::kAdsEnabledThenDisabledRewardsOff);
       prefs->SetBoolean(brave_ads::prefs::kAdsWereDisabled, true);
     }
   } else if (pref == brave_rewards::prefs::kBraveRewardsEnabled) {
-    // Rewards pref was changed
+    // Rewards pref was changed.
     if (prefs->GetBoolean(brave_ads::prefs::kAdsWereDisabled)) {
+      DCHECK(!ads_enabled);
```
Comment (masparrow) [inline components/brave_rewards/browser/rewards_p3a.cc]: See above comment from @tmancey - if we DCHECK it's best to also then fail gracefully?

File: components/brave_rewards/browser/rewards_p3a.cc [hunk 4]
```diff
@@ -34,76 +38,93 @@ void RecordAutoContributionsState(AutoContributionsP3AState state, int count) {
...
+      // TODO(ifremov): DCHECK(rewards_enabled)?
       brave_rewards::RecordAdsState(
           rewards_enabled ? AdsP3AState::kAdsEnabledThenDisabledRewardsOn :
                             AdsP3AState::kAdsEnabledThenDisabledRewardsOff);
       prefs->SetBoolean(brave_ads::prefs::kAdsWereDisabled, true);
     }
   } else if (pref == brave_rewards::prefs::kBraveRewardsEnabled) {
-    // Rewards pref was changed
+    // Rewards pref was changed.
     if (prefs->GetBoolean(brave_ads::prefs::kAdsWereDisabled)) {
```
Comment (masparrow) [inline components/brave_rewards/browser/rewards_p3a.cc]: Not even a nit: This is existing code, I know - I'm just asking this for my own info; 

`const bool ads_enabled = prefs->GetBoolean(brave_ads::prefs::kEnabled)`
&
`if (prefs->GetBoolean(brave_ads::prefs::kAdsWereDisabled) {`

Do we therefore track (elsewhere) that Ads are now Enabled, but they _were_ Disabled... ? I'm not clear on why two separate pref flags so just curious.

File: components/brave_rewards/browser/rewards_p3a.cc (hunk 3 above)
Comment (iefremov) [reply components/brave_rewards/browser/rewards_p3a.cc]: This is generally a bad practice https://chromium.googlesource.com/chromium/src/+/master/styleguide/c++/c++.md#check_dcheck_and-notreached

we should either write `DCHECK` OR fail gracefully

File: components/brave_rewards/browser/rewards_p3a.cc (hunk 4 above)
Comment (iefremov) [reply components/brave_rewards/browser/rewards_p3a.cc]: yep, correct - this is what we actually have to track
```
Q17: Have you enabled Brave Ads?
(Brave.Rewards.AdsState)

No wallet created
Wallet created, ads not enabled
Wallet created and ads enabled
Wallet created, ads enabled, then disabled, while Brave Rewards remains on
Wallet created, ads enabled, then disabled, Brave Rewards disabled
```

**Comment by iefremov** [comment]
rebased & waiting for CI

File: components/brave_rewards/browser/rewards_p3a.cc (hunk 3 above)
Comment (masparrow) [reply components/brave_rewards/browser/rewards_p3a.cc]: Ah, thanks @iefremov - I hadn't seen this, thanks for clarifying!

File: components/brave_rewards/browser/rewards_p3a.cc (hunk 4 above)
Comment (masparrow) [reply components/brave_rewards/browser/rewards_p3a.cc]: Thanks @iefremov !

**Pull request merged**